File: cache/src/strategies/lru/doublyLinkedList.js
const Node = require("./node");

class DoublyLinkedList {
    constructor(){
        this.head = new Node(null, null);
        this.tail = new Node(null, null);
        this.head.next  = this.tail;
        this.tail.prev = this.head;
    }

    addToFront(node){
        node.next = this.head.next; 
        node.next.prev = node; 
        node.prev = this.head;
        this.head.next = node; 
    }

    removeNode(node){
      node.prev.next = node.next;
      node.next.prev = node.prev;
    }

    removeLRU(){
     const lru = this.tail.prev;
     this.removeNode(lru);
     return lru;
    }
}

module.exports = DoublyLinkedList;

File: javacache/src/main/java/org/example/strategy/lru/DoublyLinkedList.java
package org.example.strategy.lru;

public class DoublyLinkedList {

    private Node head;
    private Node tail;

    public DoublyLinkedList() {
        head = new Node(-1, -1);
        tail = new Node(-1, -1);

        head.next = tail;
        tail.prev = head;
    }

    public void addToFront(Node node) {
        node.next = head.next;
        node.next.prev = node;
        head.next = node;
        node.prev = head;

    }

    public void removeNode(Node node) {
        node.prev.next = node.next;
        node.next.prev = node.prev;
    }

    public Node removeLRU() {
        Node lru = tail.prev;
        removeNode(lru);
        return lru;
    }

}
